cli/src/pcluster/templates/cw_dashboard_builder.py

```diff
@@ -0,0 +1,462 @@
+            metrics=efs_metrics, storages_list=efs_volumes_list, namespace="AWS/EFS", dimension_name="FileSystemId"
+        )
+
+        # Conditional EFS metrics
+        conditional_efs_metrics_params = [
+            GeneralMetric("Percent IO Limit", "PercentIOLimit", ["generalPurpose"]),
+        ]
+        widgets_list += self._add_conditional_storage_widgets(
+            conditional_metrics=conditional_efs_metrics_params,
+            volumes_list=efs_volumes_list,
+            namespace="AWS/EFS",
+            dimension_vol_name="FileSystemId",
+            vol_attribute_name="performance_mode",
+        )
+
+        # Add unconditional metrics graphs and conditional EFS metrics graphs
+        self.cloudwatch_dashboard.add_widgets(*widgets_list)
+        self._update_coord_after_section(self.graph_height)
+
+    def _add_fsx_metrics_graphs(self):
+        self._add_text_widget("## FSx Metrics")
+        fsx_volumes_list = self.storage_resources[SharedStorageType.FSX]
+
+        # Unconditional FSX metrics
+        fsx_metrics = [
+            GeneralMetric("Data Read/Write Ops", ["DataReadOperations", "DataWriteOperations"]),
+            GeneralMetric("Data Read/Write Bytes", ["DataReadBytes", "DataWriteBytes"]),
+            GeneralMetric("Free Data Storage Capacity", ["FreeDataStorageCapacity"]),
+        ]
+
+        # Add FSx metrics graphs and update coordinates
+        widgets_list = self._add_storage_widgets(
+            metrics=fsx_metrics,
+            storages_list=fsx_volumes_list,
+            namespace="AWS/FSX",
+            dimension_name="FileSystemId",
+        )
+        self.cloudwatch_dashboard.add_widgets(*widgets_list)
+        self._update_coord_after_section(self.graph_height)
+
+    def _add_cw_log(self):
+        # Create a text widget for subtitle "Head Node Logs"
+        self._add_text_widget("## Head Node Logs")
+
+        dcv_enabled = self.config.is_dcv_enabled
+        scheduler = self.config.scheduling.scheduler
+        base_os = self.config.image.os
+        head_private_ip = self.head_node_instance.attr_private_ip
+
+        Condition = namedtuple("Condition", ["allowed_values", "param"])
+        Filter = namedtuple("Filter", ["pattern", "param"], defaults=["@logStream"])
+        CWLogWidget = namedtuple(
+            "CWLogWidget",
+            ["title", "conditions", "fields", "filters", "sort", "limit"],
+            defaults=[None, ["@timestamp", "@message"], None, "@timestamp desc", 100],
+        )
+        SectionWidgets = namedtuple("SectionWidgets", ["section_title", "widgets"])
+
+        sections_widgets = [
+            SectionWidgets(
+                "ParallelCluster's logs",
+                [
+                    CWLogWidget(
+                        title="clustermgtd",
+                        conditions=[Condition(["slurm"], scheduler)],
+                        filters=[Filter(pattern="{0}.*clustermgtd".format(head_private_ip))],
```

Comment:
We can use fstrings everywhere rather than format. You can see [here](https://realpython.com/python-f-strings/) why.
e.g. `pattern=f"{head_private_ip}.*clustermgtd"`

Reply:
Changed